web/src/components/countrytable.js

```diff
@@ -1,649 +1,502 @@
-        })
-        .on('mouseout', function (d) {
-            if (that.exchangeMouseOutHandler)
-                that.exchangeMouseOutHandler.call(this, d.key, that._data, that._displayByEmissions);
-        })
-        .on('mousemove', function (d) {
-            if (that.exchangeMouseMoveHandler)
-                that.exchangeMouseMoveHandler.call(this, d.key, that._data, that._displayByEmissions);
-        });
-
-    gNewRow.merge(selection).select('text')
-        .text(function(d) { return d.key; });
-    const co2intensity = this._electricityMixMode === 'consumption'
-      ? this._data.co2intensity
-      : this._data.co2intensityProduction;
-    d3.select('.country-emission-intensity')
-        .text(Math.round(co2intensity) || '?');
-
-    var priceData = this._data.price || {};
-    var hasPrice = priceData.value != null;
-    d3.select('.country-spot-price')
-        .text(hasPrice ? Math.round(priceData.value) : '?')
-        .style('color', (priceData.value || 0) < 0 ? 'red' : undefined);
-    d3.select('.country-spot-price-currency')
-        .text(getSymbolFromCurrency(priceData.currency) || priceData.currency || '?')
-    d3.select('#country-emission-rect, .left-panel-zone-details .emission-rect')
-        .transition()
-        .duration(ignoreTransitions ? 0 : this.TRANSITION_DURATION)
-        .style('background-color',
-            co2intensity ?
-                that.co2color()(co2intensity) : 'gray');
-    d3.select('.country-data-source')
-        .text(this._data.source || '?');
-
-    this.resize();
-
-    return this;
+import React, {
+  useRef,
+  useMemo,
+  useEffect,
+  useState,
+} from 'react';
+import { connect } from 'react-redux';
+import { scaleLinear } from 'd3-scale';
+import { max as d3Max, min as d3Min } from 'd3-array';
+import { precisionPrefix, formatPrefix } from 'd3-format';
+import { isArray, isFinite } from 'lodash';
+
+import { dispatchApplication } from '../store';
+import { useWidthObserver } from '../effects';
+import { getCurrentZoneData, getSelectedZoneExchangeKeys } from '../selectors';
+import { getCo2Scale } from '../helpers/scales';
+import { modeOrder, modeColor } from '../helpers/constants';
+import { flagUri } from '../helpers/flags';
+import { __ } from '../helpers/translation';
+
+import CountryTableOverlayIfNoData from './countrytableoverlayifnodata';
+
+const LABEL_MAX_WIDTH = 102;
+const TEXT_ADJUST_Y = 11;
+const ROW_HEIGHT = 13;
+const PADDING_Y = 7;
+const PADDING_X = 5;
+const FLAG_SIZE = 16;
+const RECT_OPACITY = 0.8;
+const X_AXIS_HEIGHT = 15;
+const SCALE_TICKS = 4;
+
+function handleRowMouseMove(isMobile, mode, data, ev) {
+  // If in mobile mode, put the tooltip to the top of the screen for
+  // readability, otherwise float it depending on the cursor position.
+  const tooltipPosition = !isMobile
+    ? { x: ev.clientX - 7, y: ev.clientY - 7 }
+    : { x: 0, y: 0 };
+  dispatchApplication('tooltipPosition', tooltipPosition);
```

Comment:
shouldn't we dispatch them at the same time as the two next (for performance?)

Reply:
Yeah that's a good point, addressed in https://github.com/tmrowco/electricitymap-contrib/pull/2263/commits/cdc1d8b2ed0d09c040bfa31de1c0208dbfcbc245 :+1:

Reply:
great!